crates/sui-transaction-builder/src/lib.rs

```diff
@@ -400,16 +402,19 @@ impl<Mode: ExecutionMode> TransactionBuilder<Mode> {
         )?;
         let mut check_args = Vec::new();
         let mut objects = BTreeMap::new();
-        for arg in json_args {
+        for (arg, expected_type) in json_args_and_tokens {
             check_args.push(match arg {
-                SuiJsonCallArg::Object(id) => {
-                    CheckCallArg::Object(self.get_object_arg(id, &mut objects).await?)
-                }
+                SuiJsonCallArg::Object(id) => CheckCallArg::Object(
+                    self.get_object_arg(id, &mut objects, expected_type).await?,
+                ),
                 SuiJsonCallArg::Pure(p) => CheckCallArg::Pure(p),
                 SuiJsonCallArg::ObjVec(v) => {
                     let mut object_ids = vec![];
                     for id in v {
-                        object_ids.push(self.get_object_arg(id, &mut objects).await?);
+                        object_ids.push(
```

Comment:
There are some edge case here: https://mysten-labs.slack.com/archives/C04RHGZ75A4/p1678490907579999?thread_ts=1678462995.547919&cid=C04RHGZ75A4 will follow up in a separate PR

Reply:
Resolved by https://github.com/MystenLabs/sui/pull/9130